test/validator/strict_dependencies_test.dart

```diff
@@ -0,0 +1,176 @@
+      d.file(path.join(appPath, 'lib', 'library.dart'), r'''
+        import 'dart:async';
+        import 'dart:collection';
+        import 'dart:typed_data';
+      ''').create();
+      expectNoValidationError(strictDeps);
+    });
+
+    integration('imports itself', () {
+      d.file(path.join(appPath, 'lib', 'library.dart'), r'''
+        import 'package:test_pkg/test_pkg.dart';
+      ''').create();
+      expectNoValidationError(strictDeps);
+    });
+
+    integration('has a relative import', () {
+      d.file(path.join(appPath, 'lib', 'library.dart'), r'''
+        import 'some/relative/path.dart';
+      ''').create();
+      expectNoValidationError(strictDeps);
+    });
+
+    integration('has an absolute import', () {
+      d.file(path.join(appPath, 'lib', 'library.dart'), r'''
+        import 'file://shared/some/library.dart';
+      ''').create();
+      expectNoValidationError(strictDeps);
+    });
+
+
+    integration('has a parse error preventing reading directives', () {
+      d.file(path.join(appPath, 'lib', 'library.dart'), r'''
+        import not_supported_keyword 'dart:async';
+      ''').create();
+
+      expectNoValidationError(strictDeps);
+    });
+  });
+
+  group('should consider a package invalid if it', () {
+    setUp(d.validPackage.create);
+
+    integration('does not declare an "import" as a dependency', () {
+      d.file(path.join(appPath, 'lib', 'library.dart'), r'''
+        import 'package:silly_monkey/silly_monkey.dart';
+      ''').create();
+
+      expectValidationWarning(strictDeps);
+    });
+
+    integration('does not declare an "export" as a dependency', () {
+      d.file(path.join(appPath, 'lib', 'library.dart'), r'''
+        export 'package:silly_monkey/silly_monkey.dart';
+      ''').create();
+
+      expectValidationWarning(strictDeps);
+    });
+
+    integration('has an invalid URI', () {
+      d.file(path.join(appPath, 'lib', 'library.dart'), r'''
+        import 'package:/';
+      ''').create();
+
+      expectValidationWarning(strictDeps);
+    });
+
+    integration('declares an "import" as a devDependency for lib/', () {
+      d.dir(appPath, [
+        d.libPubspec("test_pkg", "1.0.0", devDeps: {
+          "silly_monkey": "^1.2.3"
+        }, sdk: ">=1.8.0 <2.0.0"),
+        d.dir('lib', [
+          d.file('library.dart', r'''
+            import 'package:silly_monkey/silly_monkey.dart';
+          '''),
+        ]),
+      ]).create();
+      expectValidationWarning(strictDeps);
+    });
+
+    integration('declares an "import" as a devDependency for bin/', () {
+      d.dir(appPath, [
+        d.libPubspec("test_pkg", "1.0.0", devDeps: {
+          "silly_monkey": "^1.2.3"
+        }, sdk: ">=1.8.0 <2.0.0"),
+        d.dir('bin', [
+          d.file('library.dart', r'''
+            import 'package:silly_monkey/silly_monkey.dart';
+          '''),
+        ]),
+      ]).create();
+      expectValidationWarning(strictDeps);
+    });
+  });
```

Comment:
Still needs no-warning tests for:

* An `export` of a dependency.
* A top level `.dart` file that contains an invalid dependency.
* A non-Dart file that looks like a Dart file that contains an invalid dependency.

And a warning test for Dart files with URIs like `"package:"`, `"package:foo"`, and `"package:/]"`.